### PR TITLE
Profile header update

### LINKED ISSUE #2445: ProfileHeader needs a rewrite
Half of user setting is done in ctor, half is done in `User_Set` with no logical separation.

Should be updated to work in the same way at `BeatmapSetOverlay`.

### PR DESCRIPTION
**Pull request by jorolf**
Updates the profile header to the new design shown on the osu website

- [x] Depends on https://github.com/ppy/osu-resources/pull/50
- Fixes #2445 (maybe? It does remove the user parameter from the ctor)

---

Updates the profile header to the new design shown on the osu website

### CODE REVIEW

File: osu.Game/Overlays/Profile/Header/ProfileHeaderTabControl.cs [hunk 1]
```diff
@@ -0,0 +1,149 @@
...
+                foreach (TabItem<string> tabItem in TabContainer)
+                {
+                    ((ProfileHeaderTabItem)tabItem).AccentColour = value;
+                }
+            }
+        }
+
+        public MarginPadding Padding
+        {
+            set => TabContainer.Padding = value;
```
Comment (UselessToucan) [inline osu.Game/Overlays/Profile/Header/ProfileHeaderTabControl.cs]: accessors order is messed up

**Review by UselessToucan — CHANGES_REQUESTED** [review]

File: osu.Game/Overlays/Profile/ProfileHeader.cs [hunk 2]
```diff
@@ -343,71 +714,149 @@ private void loadUser()
...
+            userStats.Add(new UserStatsLine("Play Count", user.Statistics.PlayCount.ToString("#,##0")));
+            userStats.Add(new UserStatsLine("Total Score", user.Statistics.TotalScore.ToString("#,##0")));
+            userStats.Add(new UserStatsLine("Total Hits", user.Statistics.TotalHits.ToString("#,##0")));
+            userStats.Add(new UserStatsLine("Maximum Combo", user.Statistics.MaxCombo.ToString("#,##0")));
+            userStats.Add(new UserStatsLine("Replays Watched by Others", user.Statistics.ReplaysWatched.ToString("#,##0")));
+
+            followerText.Text = user.FollowerCount?.Length > 0 ? user.FollowerCount[0].ToString("#,##0") : "0";
+
+            if (!user.PMFriendsOnly)
+                messageButton.Action = () =>
```
Comment (naoey) [inline osu.Game/Overlays/Profile/ProfileHeader.cs]: trying to message yourself throws an unhandled exception

File: osu.Game/Overlays/Profile/ProfileHeader.cs [hunk 3]
```diff
@@ -2,54 +2,111 @@
...
+        private readonly LinkFlowContainer bottomTopLinkContainer;
+        private readonly LinkFlowContainer bottomLinkContainer;
+
+        private const float cover_height = 150;
+        private const float cover_info_height = 75;
         private const float avatar_size = 110;
-        private const float level_position = 30;
-        private const float level_height = 60;
-        private const float stats_width = 280;
+        private static readonly Dictionary<string, string> play_styles = new Dictionary<string, string>
```
Comment (peppy) [inline osu.Game/Overlays/Profile/ProfileHeader.cs]: make this an enum instead, using `[Description("Touch Screen")]`

Comment (jorolf) [reply osu.Game/Overlays/Profile/ProfileHeader.cs]: I'd have to manually convert the json string into the enum though...

**Comment by peppy** [comment]
On a quick look, I think I would prefer this is re-structured a bit. Right now it is still using the original logic of "clear everything then make new everything". The direction we're going for is something like `BeatmapSetOverlay`, which creates each component once and then passes down the updated model (`BeatmapSet`) when it changes. Each component then handles the transition to showing the new version. Each component also handles this model going null (during a fetch from server) to go into a placeholder/loading state.

**Comment by jorolf** [comment]
Should I split the ProfileHeader into more components or should I leave it "whole". Most of the stuff already properly updates when the user property changes. And the rest mostly consists of containers that I could clear before adding new content to them.

File: osu.Game/Overlays/Profile/Header/CenterHeaderContainer.cs [hunk 4]
```diff
@@ -0,0 +1,324 @@
...
+        }
+
+        private void updateDisplay()
+        {
+            followerText.Text = user.FollowerCount?.Length > 0 ? user.FollowerCount[0].ToString("#,##0") : "0";
+
+            if (!user.PMFriendsOnly && apiAccess.LocalUser.Value.Id != user.Id)
+            {
+                messageButton.Show();
+                messageButton.Action = () =>
```
Comment (smoogipoo) [inline osu.Game/Overlays/Profile/Header/CenterHeaderContainer.cs]: Why not move all this logic inside a `MessageButton` class?

**Review by smoogipoo — DISMISSED** [review]

File: osu.Game/Overlays/Profile/Header/TopHeaderContainer.cs [hunk 5]
```diff
@@ -0,0 +1,214 @@
...
+using osu.Framework.Graphics.Shapes;
+using osu.Game.Graphics;
+using osu.Game.Graphics.Sprites;
+using osu.Game.Graphics.UserInterface;
+using osu.Game.Users;
+using osuTK;
+
+namespace osu.Game.Overlays.Profile.Header
+{
+    public class TopHeaderContainer : Container
```
[the remark this answers is not in the document]
Comment (jorolf) [reply osu.Game/Overlays/Profile/Header/TopHeaderContainer.cs]: ``TopHeaderUserDetails``? ``TopUserDetails``? ``TopUserDetailsHeader``?

File: osu.Game/Overlays/Profile/Header/SupporterIcon.cs [hunk 6]
```diff
@@ -7,58 +7,79 @@
...
             };
         }
 
+        public override bool Invalidate(Invalidation invalidation = Invalidation.All, Drawable source = null, bool shallPropagate = true)
+        {
+            bool invalid = base.Invalidate(invalidation, source, shallPropagate);
+
+            if ((invalidation & Invalidation.DrawSize) != 0)
+            {
+                iconContainer.Padding = new MarginPadding { Horizontal = DrawHeight / 2 };
```
[the remark this answers is not in the document]
Comment (jorolf) [reply osu.Game/Overlays/Profile/Header/SupporterIcon.cs]: I was going for a dynamic supporter icon size, however the FillFlowContainer didn't like that (SpriteIcon would have to be FillMode = FillMode.Fit and RelativeSizeAxes = Axes.Both)

File: osu.Game/Overlays/Profile/Header/MedalHeaderContainer.cs [hunk 7]
```diff
@@ -0,0 +1,136 @@
...
+            };
+        }
+
+        private void updateDisplay()
+        {
+            var badges = User.Badges;
+            badgeFlowContainer.Clear();
+            if (badges?.Length > 0)
+            {
+                Show();
```
[the remark this answers is not in the document]
Comment (jorolf) [reply osu.Game/Overlays/Profile/Header/MedalHeaderContainer.cs]: It hides the Container if the user doesn't have any badges

**Comment by smoogipoo** [comment]
Wanna resolve conflicts on this one? @jorolf

**Comment by smoogipoo** [comment]
This is a really solid implementation, nice work. I've just cleaned up a few things here and there to match our code style.

Requesting review @peppy .

**Review by peppy — APPROVED** [review]

**Pull request merged**